qrenderdoc/Windows/TextureViewer.cpp

```diff
@@ -3944,40 +3953,58 @@ void TextureViewer::on_cancelTextureListFilter_clicked()
   ui->textureListFilter->setCurrentText(QString());
 }
 
-void TextureViewer::on_textureListFilter_editTextChanged(const QString &text)
+void TextureViewer::on_colSelect_clicked()
 {
-  TextureListItemModel *model = (TextureListItemModel *)ui->textureList->model();
-
-  if(model == NULL)
-    return;
+  QStringList headers;
+  for(int i = 0; i < TextureListFilter::Column_Count; ++i)
+  {
+    headers.push_back(
+        ui->textureList->model()->headerData(i, Qt::Horizontal, Qt::DisplayRole).toString());
+  }
 
-  model->reset(TextureListItemModel::String, text);
+  UpdateVisibleColumns(tr("Select Texture List Columns"), TextureListFilter::Column_Count,
+                       ui->textureList->header(), headers);
 }
 
-void TextureViewer::on_textureListFilter_currentIndexChanged(int index)
+void TextureViewer::on_textureListFilter_editTextChanged(const QString &text)
 {
-  refreshTextureList();
+  refreshTextureList(FilterType::String, text);
 }
 
-void TextureViewer::refreshTextureList()
+void TextureViewer::on_textureListFilter_currentIndexChanged(int index)
 {
-  TextureListItemModel *model = (TextureListItemModel *)ui->textureList->model();
-
-  if(model == NULL)
-    return;
-
   if(ui->textureListFilter->currentIndex() == 1)
-    model->reset(TextureListItemModel::Textures, QString());
+    refreshTextureList(FilterType::Textures, QString());
   else if(ui->textureListFilter->currentIndex() == 2)
-    model->reset(TextureListItemModel::RenderTargets, QString());
+    refreshTextureList(FilterType::RenderTargets, QString());
   else
-    model->reset(TextureListItemModel::String, ui->textureListFilter->currentText());
+    refreshTextureList(FilterType::String, ui->textureListFilter->currentText());
 }
 
-void TextureViewer::on_textureList_clicked(const QModelIndex &index)
+void TextureViewer::texture_itemActivated(RDTreeWidgetItem *item, int column)
 {
-  ResourceId id = index.model()->data(index, Qt::UserRole).value<ResourceId>();
-  ViewTexture(id, false);
+  QVariant tag = item->tag();
+  if(!tag.canConvert<ResourceId>())
+    return;
+
+  TextureDescription *tex = m_Ctx.GetTexture(tag.value<ResourceId>());
+  if(!tex)
+    return;
+
+  if(tex->type == TextureType::Buffer)
+  {
+    IBufferViewer *viewer = m_Ctx.ViewTextureAsBuffer(
+        tex->resourceId, Subresource(), BufferFormatter::GetTextureFormatString(*tex));
+
+    m_Ctx.AddDockWindow(viewer->Widget(), DockReference::AddTo, this);
+  }
+  else
+  {
+    if(!m_Ctx.HasTextureViewer())
+      m_Ctx.ShowTextureViewer();
+    ITextureViewer *viewer = m_Ctx.GetTextureViewer();
+    viewer->ViewTexture(tex->resourceId, true);
```

Comment:
This show/get is not necessary here - we can call directly to `ViewTexture` from inside the texture viewer :smile: .